Fix MIPS ABI for large aggregates when passed by value

rracariu opened the pull request:
https://github.com/ldc-developers/ldc/issues/3050

kinke commented:
Thx; looks okay, although it could be made simpler. The usage of `IntegerRewrite` should be purely optional; it's used for Win64 because it's needed to match the MS specs (pass struct of power-of-2-size <= 64 bits in GP register). The magic C++ structs aren't used anymore IIRC (replaced by magic enums).

You should IMO add a clarifying comment somewhere, referencing the issue and explaining why `byval` isn't used (and that the `IndirectByvalRewrite` is just a hack to make the interface calls work, and that the ABI is nowhere near C-compatible).

Wrt. test, adding an explicit one certainly wouldn't hurt. It might be already covered by existing tests (if so, probably in dmd-testsuite), but I guess you don't know either. ;)

rracariu commented:
OK, thanks!

I simplified the code as suggested and added the comment.
As for the test, this one looks to be doing it ldc\tests\codegen\indirect_byval_rewrite.d.

kinke commented:
That test only makes sure `IndirectByvalRewrite` itself isn't too dumb/inefficient (in-place construction etc.). We need a test specifically for #3050, making sure that interface calls work with non-tiny aggregates.

---

Some aggregates are now not rewritten (those rewritable to integers), which is probably less than ideal (e.g., a struct with 8 bools is probably passed as 8 separate bools). I'll rework the logic sometime; it'd be nice if you could try to reduce the #3050 testcase to the bare minimum and verify that it still reproducibly fails on MIPS.

rracariu commented:
> That test only makes sure `IndirectByvalRewrite` itself isn't too dumb/inefficient (in-place construction etc.). We need a test specifically for #3050, making sure that interface calls work with non-tiny aggregates.
> 
OK, I'll try to simplify the code.

> Some aggregates are now not rewritten (those rewritable to integers), which is probably less than ideal (e.g., a struct with 8 bools is probably passed as 8 separate bools). I'll rework the logic etc. sometime; it'd be nice if you could try to reduce the #3050 testcase to the bare minimum and verify that it still reproducibly fails on MIPS (before this PR obviously).

If I remove that test stuff will break, I think it might be related to non-power-of-2 aggregates.

kinke commented:
> If I remove that test stuff will break, I think it might be related to non-power-of-2 aggregates.

So just using `bool passPointerToHiddenCopy(Type *t, bool) const { return isAggregate(t); }` does not work? If so, can you check whether a somewhat relaxed `TY ty = t->toBasetype()->ty; return ty == Tstruct || ty == Tsarray;` does?
